pkg/graveler/committed/apply.go

```diff
@@ -26,99 +26,129 @@ func incrementDiffSummary(d *graveler.DiffSummary, typ graveler.DiffType) {
-				return fmt.Errorf("write source record: %w", err)
+			if err := writer.WriteRecord(*iterValue); err != nil {
+				return 0, fmt.Errorf("write iter record: %w", err)
 			}
-			if !source.Next() {
+			count++
+			if !iter.Next() {
 				break
 			}
 		}
 	}
-	return source.Err()
+	return count, iter.Err()
 }
 
-// applyFromDiffs applies all changes from diffs to writer and returns the number elements it
-// added.
-func applyFromDiffs(ctx context.Context, logger logging.Logger, writer MetaRangeWriter, diffs graveler.ValueIterator) (int, error) {
-	numAdded := 0
-	for {
-		select {
-		case <-ctx.Done():
-			return 0, ctx.Err()
-		default:
-		}
-
-		diffValue, haveDiffs := diffs.Value(), diffs.Next()
-		if diffValue.IsTombstone() {
-			// internal error but no data lost: deletion requested of a
-			// file that was not there.
-			logger.WithField("id", string(diffValue.Identity)).Warn("[I] unmatched delete")
-			continue
-		}
-		if logger.IsTracing() {
-			logger.WithFields(logging.Fields{
-				"key":       string(diffValue.Key),
-				"ID":        string(diffValue.Identity),
-				"tombstone": diffValue.IsTombstone(),
-			}).Trace("write key from diffs at end")
-		}
-		if err := writer.WriteRecord(*diffValue); err != nil {
-			return 0, fmt.Errorf("write added record: %w", err)
-		}
-		numAdded++
-		if !haveDiffs {
-			break
+func hasChanges(summary graveler.DiffSummary) bool {
+	for _, changes := range summary.Count {
+		if changes > 0 {
+			return true
 		}
 	}
-	return numAdded, nil
+	return false
 }
 
-func Apply(ctx context.Context, writer MetaRangeWriter, source Iterator, diffs graveler.ValueIterator, opts *ApplyOptions) (graveler.DiffSummary, error) {
+func Apply(ctx context.Context, writer MetaRangeWriter, source Iterator, diffs Iterator, opts *ApplyOptions) (graveler.DiffSummary, error) {
 	logger := logging.FromContext(ctx)
 	ret := graveler.DiffSummary{Count: make(map[graveler.DiffType]int)}
 	haveSource, haveDiffs := source.Next(), diffs.Next()
-	changed := false
 	for haveSource && haveDiffs {
 		select {
 		case <-ctx.Done():
 			return ret, ctx.Err()
 		default:
 		}
-
 		sourceValue, sourceRange := source.Value()
-		diffValue := diffs.Value()
-		if sourceValue == nil {
+		diffValue, diffRange := diffs.Value()
+		switch {
+		case diffValue == nil && sourceValue == nil:
+			switch {
+			case bytes.Compare(diffRange.MaxKey, sourceRange.MinKey) < 0:
+				// insert diff
+				writer.WriteRange(*diffRange)
+				addIntoDiffSummary(&ret, graveler.DiffTypeAdded, int(diffRange.Count))
+				haveDiffs = diffs.Next()
+			case bytes.Compare(sourceRange.MaxKey, diffRange.MinKey) < 0:
+				//insert source
+				writer.WriteRange(*sourceRange)
+				haveSource = source.Next()
+			case diffRange.ID == sourceRange.ID && diffRange.IsTombstone():
+				addIntoDiffSummary(&ret, graveler.DiffTypeRemoved, int(diffRange.Count))
+				haveSource = source.Next()
+				haveDiffs = diffs.Next()
+			default:
```

Comment:
:point_right: I don't understand: are these all the options?  What about when `sourceRange` and `diffRange` have overlapping keys?

```
source               [min].....[max]
        ------------------------------------------------------------------------------------------------
diffs       [min...........max]

Reply:
In case of overlapping keys, we call next on both ranges In order to compare keys.